src/server/v2.0/handler/artifact.go

```diff
@@ -16,32 +16,157 @@ package handler
+		proMgr:  project.Mgr,
+		repoMgr: repository.Mgr,
+	}
+}
+
+type artifactAPI struct {
 	BaseAPI
+	artCtl  artifact.Controller
+	proMgr  project.Manager
+	repoMgr repository.Manager
 }
 
-// DeleteArtifact ...
-func (api *ArtifactAPI) DeleteArtifact(ctx context.Context, params operation.DeleteArtifactParams) middleware.Responder {
-	return operation.NewDeleteArtifactOK()
+// TODO do auth in a separate middleware
+
+func (a *artifactAPI) ListArtifacts(ctx context.Context, params operation.ListArtifactsParams) middleware.Responder {
+	// set query
+	query := &q.Query{
+		Keywords: map[string]interface{}{},
+	}
+	if params.Type != nil {
+		query.Keywords["Type"] = *(params.Type)
+	}
+	if params.Page != nil {
+		query.PageNumber = *(params.Page)
+	}
+	if params.PageSize != nil {
+		query.PageSize = *(params.PageSize)
+	}
+	repository, err := a.repoMgr.GetByName(ctx, fmt.Sprintf("%s/%s", params.ProjectName, params.RepositoryName))
+	if err != nil {
+		return a.SendError(ctx, err)
+	}
+	query.Keywords["RepositoryID"] = repository.RepositoryID
+
+	// set option
+	option := &artifact.Option{
+		WithTag: true, // return the tag by default
+	}
+	if params.WithTag != nil {
+		option.WithTag = *(params.WithTag)
+	}
+	if option.WithTag {
+		if params.WithImmutableStatus != nil {
+			option.TagOption = &artifact.TagOption{
+				WithImmutableStatus: *(params.WithImmutableStatus),
+			}
+		}
+	}
+	if params.WithLabel != nil {
+		option.WithLabel = *(params.WithLabel)
+	}
+	if params.WithScanReport != nil {
+		option.WithScanReport = *(params.WithScanReport)
+	}
+	if params.WithSignatrue != nil {
+		option.WithSignature = *(params.WithSignatrue)
+	}
+
+	// list artifacts according to the query and option
+	total, arts, err := a.artCtl.List(ctx, query, option)
+	if err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	var artifacts []*models.Artifact
+	for _, art := range arts {
+		artifacts = append(artifacts, convertArtifact(art))
+	}
+
+	// TODO add link header
+	return operation.NewListArtifactsOK().WithXTotalCount(total).WithLink("").WithPayload(artifacts)
 }
 
-// ListArtifacts ...
-func (api *ArtifactAPI) ListArtifacts(ctx context.Context, params operation.ListArtifactsParams) middleware.Responder {
-	return operation.NewListArtifactsOK()
+func (a *artifactAPI) GetArtifact(ctx context.Context, params operation.GetArtifactParams) middleware.Responder {
+	// TODO implement
+	return operation.NewGetArtifactOK()
 }
 
-// ReadArtifact ...
-func (api *ArtifactAPI) ReadArtifact(ctx context.Context, params operation.ReadArtifactParams) middleware.Responder {
-	return operation.NewReadArtifactOK()
+func (a *artifactAPI) DeleteArtifact(ctx context.Context, params operation.DeleteArtifactParams) middleware.Responder {
+	// TODO implement
+	return operation.NewDeleteArtifactOK()
 }
 
-// NewArtifactAPI returns API of artifacts
-func NewArtifactAPI() *ArtifactAPI {
-	return &ArtifactAPI{}
+func convertArtifact(art *artifact.Artifact) *models.Artifact {
```

Comment:
Did I remember this function is defined as a method of the API model?

Reply:
This is a different one, it converts the api model to the one generated by swagger